KAFKA-14365: Refactor Fetcher to allow different implementations

kirktrue opened the pull request:
This change refactors the existing `Fetcher` by splitting out the parts related to offset listing into a dedicated class named `OffsetsFinder`. This allows the `Fetcher` implementation to be more dedicated and thus easier to refactor (in coming commits).

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

kirktrue committed:
KAFKA-14365: Refactor Fetcher to allow different implementations

This change refactors the existing Fetcher by splitting out the parts
related to offset listing into a dedicated class named OffsetsFinder.
This allows the Fetcher implementation to be more dedicated and thus
easier to refactor (in coming commits).

kirktrue committed:
KAFKA-14365: Refactor Fetcher to allow different implementations

This change refactors the existing Fetcher by splitting out the parts
related to offset listing into a dedicated class named OffsetsFinder.
This allows the Fetcher implementation to be more dedicated and thus
easier to refactor (in coming commits).

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsFinder.java:
```diff
@@ -0,0 +1,805 @@
+     * @param timer Timer bounding how long this method can block
+     * @return The map of topics with their partition information
+     */
+    public Map<String, List<PartitionInfo>> getTopicMetadata(MetadataRequest.Builder request, Timer timer) {
```
This doesn't seem related to finding offsets. Should it be somewhere else?

kirktrue replied:
Agreed. Is adding it to `KafkaConsumer` directly OK?

hachikuji replied:
We can generalize the name perhaps? Instead of `OffsetFinder`, perhaps we can call it `MetadataManager` or something generic like that.

kirktrue replied:
Name has been changed on the latest commit.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsFinder.java:
```diff
@@ -0,0 +1,805 @@
+    private final long retryBackoffMs;
+    private final long requestTimeoutMs;
+    private final ConsumerMetadata metadata;
+    private final SubscriptionState subscriptions;
```
To get a real benefit out of this refactor, we need to reduce the dependence on some of these components. Rather than pulling state directly from `SubscriptionState`, for example, it would be better to have methods which accept the state we need as arguments. Then we can focus this component on the functional logic, such as building requests, checking errors, and handling backoff. We also, want to drop the dependence on `ConsumerNetworkClient`. All the sending/polling will be done in the background thread.

kirktrue replied:
Agreed!

Splitting the functionality was introduced as a stand-alone PR in an attempt to have more and smaller PRs vs. fewer and larger PRs. This was to be just the first PR of several.

Would you prefer to have everything fleshed out completely?

jsancio committed:
MINOR; Improve high watermark log messages (#12975)

While debugging KRaft and the metadata state machines it is helpful to always log the first time the replica discovers the high watermark. All other updates to the high watermark are logged at trace because they are more frequent and less useful.

Reviewers: Luke Chen <showuon@gmail.com>

dajac committed:
MINOR: Small refactor in KafkaApis.handleHeartbeatRequest (#12978)

This is a small follow-up to https://github.com/apache/kafka/pull/12848. 

Reviewers: Justine Olshan <jolshan@confluent.io>, Jason Gustafson <jason@confluent.io>

dajac committed:
KAFKA-14367; Add `DescribeGroups` to the new `GroupCoordinator` interface (#12855)

This patch adds `describeGroups` to the new `GroupCoordinator` interface and updates `KafkaApis` to use it.

Reviewers: Justine Olshan <jolshan@confluent.io>, Jason Gustafson <jason@confluent.io>

aglicacha committed:
KAFKA-14285: Delete quota node in zookeeper when configs are empty (#12729)

The PR resolve issue KAFKA-14285. After doing changeConfigs, check and clean quota nodes if configs are empty, to avoid infinite increasement of quota nodes in zookeeper

Reviewers: Luke Chen <showuon@gmail.com>, Igor Soarez <soarez@apple.com>, dengziming <dengziming1993@gmail.com>

gharris1727 committed:
KAFKA-13881: Add Connect package infos (#12937)

Reviewers: Chris Egerton <chrise@aiven.io>

vamossagar12 committed:
KAFKA-13602: Remove unwanted logging in RecordCollectorImpl.java (#12985)

There is unwanted logging introduced by #12803 as pointed out in this comment: #12803 (comment). This PR removes it.

Reviewers: Lucas Brutschy <lbrutschy@confluent.io>, Bruno Cadonna <cadonna@apache.org>

hachikuji committed:
MINOR: Pass snapshot ID directly in `RaftClient.createSnapshot` (#12981)

Let `RaftClient.createSnapshot` take the snapshotId directly instead of the committed offset/epoch (which may not exist). 

Reviewers: José Armando García Sancio <jsancio@apache.org>

mumrah committed:
KAFKA-14448 Let ZK brokers register with KRaft controller (#12965)

Prior to starting a KIP-866 migration, the ZK brokers must register themselves with the active
KRaft controller. The controller waits for all brokers to register in order to verify that all the
brokers can

A) Communicate with the quorum
B) Have the migration config enabled
C) Have the proper IBP set

This patch uses the new isMigratingZkBroker field in BrokerRegistrationRequest and
RegisterBrokerRecord. The type was changed from int8 to bool for BrokerRegistrationRequest (a
mistake from #12860). The ZK brokers use the existing BrokerLifecycleManager class to register and
heartbeat with the controllers.

Reviewers: Mickael Maison <mickael.maison@gmail.com>, Colin P. McCabe <cmccabe@apache.org>

kirktrue committed:
Merge branch 'trunk' into KAFKA-14365-refactor-fetcher-take-3

kirktrue committed:
Renaming OffsetsFinder to MetadataManager

kirktrue committed:
Renaming OffsetsFinder to MetadataManager

artemlivshits committed:
MINOR: Fix FetchFromFollowerIntegrationTest.testFetchFromFollowerWithRoll (#12987)

The test was added with a fix to KAFKA-14379, the problem was that the replication factor for the offset topic was 1 and consumer group coordinator got unavailable when one of the brokers got shut down.

Reviewers: David Jacot <djacot@confluent.io>

mumrah committed:
MINOR: Change KRaft ZK controller registration algorithm (#12973)

Increment the value in "/controller_epoch" when registering a KRaft controller as the active controller. Use the "kraftControllerEpoch" stored under "/controller" to ensure we are registering a newer KRaft controller.

Reviewers: Colin P. McCabe <cmccabe@apache.org>

lihaosky committed:
KAFKA-14395: add config to configure client supplier (#12944)

Implements KIP-884.

Reviewers: Matthias J. Sax <matthias@confluent.io>

mjsax committed:
MINOR: update Streams upgrade guide for 3.1 release (#12926)

Reviewers: John Roesler <john@confluent.io>, Bill Bejeck <bill@confluent.io>

dajac committed:
KAFKA-14367; Add `DeleteGroups` to the new `GroupCoordinator` interface (#12858)

This patch adds `deleteGroups` to the new `GroupCoordinator` interface and updates `KafkaApis` to use it.

Reviewers: Omnia G H Ibrahim <o.g.h.ibrahim@gmail.com>, Justine Olshan <jolshan@confluent.io>, Jason Gustafson <jason@confluent.io>

gharris1727 committed:
KAFKA-13881: Add Clients package description javadocs (#12895)


Reviewers: Mickael Maison <mickael.maison@gmail.com>  , Tom Bentley <tbentley@redhat.com>

C0urante committed:
KAFKA-13709: Add docs for exactly-once support in Connect (#12941)

Reviewers: Mickael Maison <mickael.maison@gmail.com>

akhileshchg committed:
KAFKA-14446: API forwarding support from zkBrokers to the Controller (#12961)

This PR enables brokers which are upgrading from ZK mode to KRaft mode to forward certain metadata
change requests to the controller instead of applying them directly through ZK. To faciliate this,
we now support EnvelopeRequest on zkBrokers (instead of only on KRaft nodes.)

In BrokerToControllerChannelManager, we can now reinitialize our NetworkClient. This is needed to
handle the case when we transition from forwarding requests to a ZK-based broker over the
inter-broker listener, to forwarding requests to a quorum node over the controller listener.

In MetadataCache.scala, distinguish between KRaft and ZK controller nodes with a new type,
CachedControllerId.

In LeaderAndIsrRequest, StopReplicaRequest, and UpdateMetadataRequest, switch from sending both a
zk and a KRaft controller ID to sending a single controller ID plus a boolean to express whether it
is KRaft. The previous scheme was ambiguous as to whether the system was in KRaft or ZK mode when
both IDs were -1 (although this case is unlikely to come up in practice). The new scheme avoids
this ambiguity and is simpler to understand.

Reviewers: dengziming <dengziming1993@gmail.com>, David Arthur <mumrah@gmail.com>, Colin P. McCabe <cmccabe@apache.org>

vamossagar12 committed:
Removing Multicasting partitioner for IQ (#12977)

Follow up PR for KIP-837. We don't want to allow multicasting for IQ. This PR imposes that restriction.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>

cmccabe committed:
MINOR: ControllerServer should use the new metadata loader and snapshot generator (#12983)

This PR introduces the new metadata loader and snapshot generator. For the time being, they are
only used by the controller, but a PR for the broker will come soon.

The new metadata loader supports adding and removing publishers dynamically. (In contrast, the old
loader only supported adding a single publisher.) It also passes along more information about each
new image that is published. This information can be found in the LogDeltaManifest and
SnapshotManifest classes.

The new snapshot generator replaces the previous logic for generating snapshots in
QuorumController.java and associated classes. The new generator is intended to be shared between
the broker and the controller, so it is decoupled from both.

There are a few small changes to the old snapshot generator in this PR. Specifically, we move the
batch processing time and batch size metrics out of BrokerMetadataListener.scala and into
BrokerServerMetrics.scala.

Finally, fix a case where we are using 'is' rather than '==' for a numeric comparison in
snapshot_test.py.

Reviewers: David Arthur <mumrah@gmail.com>

Scanteianu committed:
MINOR: Document Offset and Partition 0-indexing, fix typo (#12753)

Add comments to clarify that both offsets and partitions are 0-indexed, and fix a minor typo. Clarify which offset will be retrieved by poll() after seek() is used in various circumstances. Also added integration tests.

Reviewers: Luke Chen <showuon@gmail.com>

kirktrue committed:
KAFKA-14496: Wrong Base64 encoder used by OIDC OAuthBearerLoginCallbackHandler (#13000)

The OAuth code to generate the Authentication header was incorrectly
using the URL-safe base64 encoder. For client IDs and/or secrets with
dashes and/or plus signs would not be encoded correctly, leading to the
OAuth server to reject the credentials.

This change uses the correct base64 encoder, per RFC-7617.

Co-authored-by: Endre Vig <vendre@gmail.com>

satishd committed:
[KAFKA-13369] Follower fetch protocol changes for tiered storage. (#11390)

This PR implements the follower fetch protocol as mentioned in KIP-405.

Added a new version for ListOffsets protocol to receive local log start offset on the leader replica. This is used by follower replicas to find the local log star offset on the leader.

Added a new version for FetchRequest protocol to receive OffsetMovedToTieredStorageException error. This is part of the enhanced fetch protocol as described in KIP-405.

We introduced a new field locaLogStartOffset to maintain the log start offset in the local logs. Existing logStartOffset will continue to be the log start offset of the effective log that includes the segments in remote storage.

When a follower receives OffsetMovedToTieredStorage, then it tries to build the required state from the leader and remote storage so that it can be ready to move to fetch state.

Introduced RemoteLogManager which is responsible for

initializing RemoteStorageManager and RemoteLogMetadataManager instances.
receives any leader and follower replica events and partition stop events and act on them
also provides APIs to fetch indexes, metadata about remote log segments.
Followup PRs will add more functionality like copying segments to tiered storage, retention checks to clean local and remote log segments. This will change the local log start offset and make sure the follower fetch protocol works fine for several cases.

You can look at the detailed protocol changes in KIP: https://cwiki.apache.org/confluence/display/KAFKA/KIP-405%3A+Kafka+Tiered+Storage#KIP405:KafkaTieredStorage-FollowerReplication

Co-authors: satishd@apache.org, kamal.chandraprakash@gmail.com, yingz@uber.com

Reviewers: Kowshik Prakasam <kprakasam@confluent.io>, Cong Ding <cong@ccding.com>, Tirtha Chatterjee <tirtha.p.chatterjee@gmail.com>, Yaodong Yang <yangyaodong88@gmail.com>, Divij Vaidya <diviv@amazon.com>, Luke Chen <showuon@gmail.com>, Jun Rao <junrao@gmail.com>

ijuma committed:
Move IndexEntry and related to storage module (#12993)

For broader context on this change, please check:

* KAFKA-14470: Move log layer to storage module

Reviewers: dengziming <dengziming1993@gmail.com>

ijuma committed:
MINOR: Use INFO logging for tools and trogdor tests (#13006)

`TRACE` is too noisy and makes the build slower.

Reviewers: David Jacot <djacot@confluent.io>

ijuma committed:
MINOR: Fix compiler errors in `RemoteIndex*` related to `OffsetPosition` (#13008)

#11390 and #12993 were merged in relatively quick succession, which resulted in compiler errors that weren't present when each change was on top of trunk.

Reviewers: Mickael Maison <mickael.maison@gmail.com>

dajac committed:
MINOR: Serialize response in KafkaApisTest (#12865)

Every request used in KafkaApiTests is serialized before hitting KafkaApis. This ensures that backward compatibility is handled correctly, especially ignorable fields. We don't do the same for the responses though. This patch fixes this and refactor how we get back the response in tests.

Reviewers: Mickael Maison <mickael.maison@gmail.com>

satishd committed:
KAFKA-14466: Move ClassloaderAwareRemoteStorageManager to storage module (#13013)

Reviewers: Ismael Juma <ismael@juma.me.uk>

jolshan committed:
KAFKA-14417: Address incompatible error code returned by broker from `InitProducerId` (#12968)

Older clients can not handle the `REQUEST_TIMED_OUT` error that is returned from `InitProducerId` when the next producerId block cannot be fetched from the controller. In this patch, we return `COORDINATOR_LOAD_IN_PROGRESS` instead which is retriable.

Reviewers: Jason Gustafson <jason@confluent.io>

philipnee committed:
KAFKA-14264; New logic to discover group coordinator (#12862)

[KAFKA-14264](https://issues.apache.org/jira/browse/KAFKA-14264)
In this patch, we refactored the existing FindCoordinator mechanism. In particular, we first centralize all of the network operation (send, poll) in `NetworkClientDelegate`, then we introduced a RequestManager interface that is responsible to handle the timing of different kind of requests, based on the implementation.  In this path, we implemented a `CoordinatorRequestManager` which determines when to create an `UnsentRequest` upon polling the request manager.

Reviewers: Jason Gustafson <jason@confluent.io>

bbejeck committed:
MINOR: No error with zero results state query (#13002)

This PR updates StateQueryResult.getOnlyPartitionResult() to not throw an IllegaArgumentException when there are 0 query results.

Added a test that will fail without this patch

Reviewers: John Roesler<vvcephei@apache.org>

mumrah committed:
KAFKA-14529 Use the MetadataVersion from ClusterConfig in ZK tests (#13020)

Reviewers: Ismael Juma <ismael@juma.me.uk>

ijuma committed:
KAFKA-14472: Move TransactionIndex and related to storage module (#12996)

For broader context on this change, please check:

* KAFKA-14470: Move log layer to storage module

Reviewers: Jun Rao <junrao@gmail.com>, Satish Duggana <satishd@apache.org>

ijuma committed:
MINOR: Use `withClassLoader` in a couple more places (#13018)

Instead of try/finally.

Reviewers: Ron Dagostino <rndgstn@gmail.com>

ijuma committed:
KAFKA-14473: Move AbstractIndex to storage module (#13007)

For broader context on this change, please check:

* KAFKA-14470: Move log layer to storage module

Reviewers: Jun Rao <junrao@gmail.com>, Satish Duggana <satishd@apache.org>

lucasbru committed:
KAFKA-14343: Upgrade tests for state updater  (#12801)

A test that verifies the upgrade from a version of Streams with
state updater disabled to a version with state updater enabled
and vice versa, so that we can offer a save upgrade path.

 - upgrade test from a version of Streams with state updater
disabled to a version with state updater enabled
 - downgrade test from a version of Streams with state updater
 enabled to a version with state updater disabled

Reviewer: Bruno Cadonna <cadonna@apache.org>

mimaison committed:
MINOR: Don't throw if MirrorMaker topics already exist (#13005)


Reviewers: Luke Chen <showuon@gmail.com>

mumrah committed:
KAFKA-14531 Fix controller snapshot interval (#13019)

This patch applies the necessary conversion of milliseconds to nanoseconds for SnapshotGenerator. 

Reviewers: Colin P. McCabe <cmccabe@apache.org>, José Armando García Sancio <jsancio@gmail.com>

kirktrue committed:
Merge branch 'KAFKA-14365-refactor-fetcher-take-3' of github.com:kirktrue/kafka into KAFKA-14365-refactor-fetcher-take-3
